protobuf/module_msg.proto

```diff
@@ -6,385 +6,869 @@ package bess.pb;
+ * between a specified min and max values.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 1
+ */
 message RandomUpdateArg {
+  /**
+   * RandomUpdate's Field specifies where to rewrite, and what values to rewrite
+   * in each packet processed.
+   */
   message Field {
-    int64 offset = 1;
-    uint64 size = 2;
-    uint64 min = 3;
-    uint64 max = 4;
+    int64 offset = 1; /// Offset in bytes for where to rewrite.
+    uint64 size = 2; /// The number of bytes to write.
+    uint64 min = 3; /// The minimum value to insert into the packet.
+    uint64 max = 4; /// The maximum value to insert into the packet.
   }
-  repeated Field fields = 1;
+  repeated Field fields = 1; /// A list of Random Update Fields.
 }
 
+/**
+ * The Rewrite module replaces an entire packet body with a packet "template"
+ * converting all packets that pass through to copies of the of one of
+ * the templates.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 1
+ */
 message RewriteArg {
-  repeated bytes templates = 1;
+  repeated bytes templates = 1; /// A list of bytestrings representing packet templates.
 }
 
+/**
+ * The RoundRobin module has a function `setGates(...)` which changes
+ * the total number of gates in the module.
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: many (configurable)
+ */
 message RoundRobinCommandSetGatesArg {
-  repeated int64 gates = 1;
+  repeated int64 gates = 1; /// A list of gate numbers to round-robin the traffic over.
 }
 
+/**
+ * The RoundRobin module has a function `setMode(...)` which specifies whether
+ * to balance traffic across gates per-packet or per-batch.
+ */
 message RoundRobinCommandSetModeArg {
-  string mode = 1;
+  string mode = 1; /// whether to perform 'packet' or 'batch' round robin partitioning.
 }
 
+/**
+ * The RoundRobin module splits packets from one input gate across multiple output
+ * gates.
+ */
 message RoundRobinArg {
-  repeated int64 gates = 1;
-  string mode = 2;
-}
-
+  repeated int64 gates = 1; /// A list of gate numbers to split packets across.
+  string mode = 2; /// Whether to split across gate with every 'packet' or every 'batch'.
+}
+
+/**
+ * The SetMetadata module adds metadata attributes to packets, which are not stored
+ * or sent out with packet data. For examples of SetMetadata use, see
+ * [`bess/bessctl/conf/attr_match.bess`](https://github.com/NetSys/bess/blob/master/bessctl/conf/metadata/attr_match.bess)
+ *
+ * __Input Gates__: 1
+ * __Output Gates__: 1
+ */
 message SetMetadataArg {
+  /**
+   * SetMetadata Attribute describes a metadata attribute and value to attach to every packet.
+   */
   message Attribute {
-    string name = 1;
-    uint64 size = 2;
+    string name = 1; /// The metadata attribute name.
+    uint64 size = 2; /// The size of values stored in this attribute.
     oneof value {
-      uint64 value_int = 3;
-      bytes value_bin = 4;
+      uint64 value_int = 3; /// An integer value to store in the packet.
```

Comment:
What byte order is the field vs. how is it stored in the packet?  I assume both are host byte order, which is fine, except that this will have curious interactions with modules that apply metadata to packets.